Wallabag for Android: unable to "Download PDF"

_From @returntrip on May 11, 2017 13:14_

### Issue details

Articles are not being downloaded as PDF in the Android app.

### Environment

* wallabag version (or git revision) that exhibits the issue: Android - ver 2.0.0b1
* How did you install wallabag? Via `git clone` or by downloading the package? Google Play
* Last wallabag version that did not exhibit the issue (if applicable): Not tested in other versions
* php version: NA
* OS: Android 7.1.2 stock on Nexus 6P
* type of hosting (shared or dedicated): shared
* which storage system you choose at install (SQLite, MySQL/MariaDB or PostgreSQL):  NA

### Steps to reproduce/test case
On the Android app:

1) Open an article
2) From the drop down menu select: "Download PDF"
3) A progress bar will sit in the task bar saying: "wallabg downloads your article [article name]"
4) The article is not downloaded


_Copied from original issue: wallabag/wallabag#3109_

Yes, i can reproduce it:

```
05-12 17:46:26.243 8428-8491/fr.gaulupeau.apps.InThePoche.debug D/EventProcessor: onDownloadFileFinishedEvent() started
05-12 17:46:26.246 8428-8491/fr.gaulupeau.apps.InThePoche.debug E/AndroidRuntime: FATAL EXCEPTION: IntentService[SecondaryService]
                                                                                  Process: fr.gaulupeau.apps.InThePoche.debug, PID: 8428
                                                                                  org.greenrobot.eventbus.EventBusException: Invoking subscriber failed
                                                                                      at org.greenrobot.eventbus.EventBus.handleSubscriberException(EventBus.java:505)
                                                                                      at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:487)
                                                                                      at org.greenrobot.eventbus.EventBus.postToSubscription(EventBus.java:416)
                                                                                      at org.greenrobot.eventbus.EventBus.postSingleEventForEventType(EventBus.java:397)
                                                                                      at org.greenrobot.eventbus.EventBus.postSingleEvent(EventBus.java:370)
                                                                                      at org.greenrobot.eventbus.EventBus.post(EventBus.java:251)
                                                                                      at fr.gaulupeau.apps.Poche.events.EventHelper.postEvent(EventHelper.java:10)
                                                                                      at fr.gaulupeau.apps.Poche.service.SecondaryService.downloadAsFile(SecondaryService.java:122)
                                                                                      at fr.gaulupeau.apps.Poche.service.SecondaryService.onHandleIntent(SecondaryService.java:68)
                                                                                      at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:67)
                                                                                      at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                      at android.os.Looper.loop(Looper.java:241)
                                                                                      at android.os.HandlerThread.run(HandlerThread.java:61)
                                                                                   Caused by: android.os.FileUriExposedException: file:///storage/emulated/0/Android/data/fr.gaulupeau.apps.InThePoche.debug/files/Open_VPN_behebt_nach_Audits_Sicherheitsm_ngel___Linux-Magazin.pdf exposed beyond app through Intent.getData()
                                                                                      at android.os.StrictMode.onFileUriExposed(StrictMode.java:1799)
                                                                                      at android.net.Uri.checkFileUriExposed(Uri.java:2351)
                                                                                      at android.content.Intent.prepareToLeaveProcess(Intent.java:9011)
                                                                                      at android.content.Intent.prepareToLeaveProcess(Intent.java:8972)
                                                                                      at android.app.PendingIntent.getActivity(PendingIntent.java:340)
                                                                                      at android.app.PendingIntent.getActivity(PendingIntent.java:302)
                                                                                      at fr.gaulupeau.apps.Poche.events.EventProcessor.onDownloadFileFinishedEvent(EventProcessor.java:373)
                                                                                      at java.lang.reflect.Method.invoke(Native Method)
                                                                                      at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:485)
                                                                                      at org.greenrobot.eventbus.EventBus.postToSubscription(EventBus.java:416) 
                                                                                      at org.greenrobot.eventbus.EventBus.postSingleEventForEventType(EventBus.java:397) 
                                                                                      at org.greenrobot.eventbus.EventBus.postSingleEvent(EventBus.java:370) 
                                                                                      at org.greenrobot.eventbus.EventBus.post(EventBus.java:251) 
                                                                                      at fr.gaulupeau.apps.Poche.events.EventHelper.postEvent(EventHelper.java:10) 
                                                                                      at fr.gaulupeau.apps.Poche.service.SecondaryService.downloadAsFile(SecondaryService.java:122) 
                                                                                      at fr.gaulupeau.apps.Poche.service.SecondaryService.onHandleIntent(SecondaryService.java:68) 
                                                                                      at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:67) 
                                                                                      at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                                      at android.os.Looper.loop(Looper.java:241) 
                                                                                      at android.os.HandlerThread.run(HandlerThread.java:61) 
05-12 17:46:26.511 8428-8428/fr.gaulupeau.apps.InThePoche.debug I/Timeline: Timeline: Activity_idle id: android.os.BinderProxy@316f4b1 time:137929945
05-12 17:46:26.513 8428-8428/fr.gaulupeau.apps.InThePoche.debug V/ReadArticleActivity: getReadingPosition() yOffset: 0, viewHeight: 1024, totalHeight: 4244
```

i am goint to test di72nn's fix now and will report back in the PR

--- a/app/src/main/AndroidManifest.xml
+++ b/app/src/main/AndroidManifest.xml
@@ -147,6 +147,15 @@
             android:name="fr.gaulupeau.apps.Poche.service.SecondaryService"
             android:exported="false" />
 
+        <provider
+            android:name="android.support.v4.content.FileProvider"
+            android:authorities="${applicationId}.fileprovider"
+            android:exported="false"
+            android:grantUriPermissions="true">
+            <meta-data
+                android:name="android.support.FILE_PROVIDER_PATHS"
+                android:resource="@xml/provider_paths"/>
+        </provider>
     </application>
 
 </manifest>

--- a/app/src/main/java/fr/gaulupeau/apps/Poche/events/EventProcessor.java
+++ b/app/src/main/java/fr/gaulupeau/apps/Poche/events/EventProcessor.java
@@ -9,6 +9,7 @@ import android.net.Uri;
 import android.os.Build;
 import android.os.Handler;
 import android.support.v4.app.NotificationCompat;
+import android.support.v4.content.FileProvider;
 import android.util.Log;
 import android.webkit.MimeTypeMap;
 import android.widget.Toast;
@@ -20,6 +21,7 @@ import java.util.Collections;
 import java.util.EnumSet;
 import java.util.Locale;
 
+import fr.gaulupeau.apps.InThePoche.BuildConfig;
 import fr.gaulupeau.apps.InThePoche.R;
 import fr.gaulupeau.apps.Poche.data.Settings;
 import fr.gaulupeau.apps.Poche.network.Updater;
@@ -361,14 +363,17 @@ public class EventProcessor {
 
         ActionResult result = event.getResult();
         if(result == null || result.isSuccess()) {
+            Context context = getContext();
+
             Intent intent = new Intent();
             intent.setAction(android.content.Intent.ACTION_VIEW);
-            Uri uri = Uri.fromFile(event.getFile());
+            Uri uri = FileProvider.getUriForFile(context,
+                    BuildConfig.APPLICATION_ID + ".fileprovider",
+                    event.getFile());
             String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(
                     event.getRequest().getDownloadFormat().toString().toLowerCase(Locale.US));
             intent.setDataAndType(uri, mimeType);
-
-            Context context = getContext();
+            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
 
             PendingIntent contentIntent = PendingIntent.getActivity(context, 0, intent, 0);
 

--- a/app/src/main/res/xml/provider_paths.xml
+++ b/app/src/main/res/xml/provider_paths.xml
@@ -1,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<paths xmlns:android="http://schemas.android.com/apk/res/android">
+    <external-files-path name="wallabag_files"/>
+</paths>

